system/lib/wasmfs/wasmfs.cpp

```diff
@@ -8,60 +8,258 @@
 
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
+#include <mutex>
 #include <stdlib.h>
+#include <utility>
 #include <vector>
 #include <wasi/api.h>
 
 extern "C" {
 
+using block_t = uint32_t;
+
 static std::vector<char> fd_write_stdstream_buffer;
 
-__wasi_errno_t __wasi_fd_write(
-  __wasi_fd_t fd, const __wasi_ciovec_t* iovs, size_t iovs_len, __wasi_size_t* nwritten) {
-  // FD 1 = STDOUT and FD 2 = STDERR.
-  // Temporary hardcoding of filedescriptor values.
+class File {
+  std::string name;
+  uint32_t uid;  // User ID of the owner
+  uint32_t gid;  // Group ID of the owning group
+  uint32_t mode; // r/w/x modes
+  time_t ctime;  // Time when the inode was last modified
+  time_t mtime;  // Time when the content was last modified
+  time_t atime;  // Time when the content was last accessed
+
+  std::vector<std::shared_ptr<block_t>> blocks; // pointers to data blocks
+
+  std::mutex file_mutex;
```

Comment:
No need to repeat the enclosing class name here:
```suggestion
  std::mutex mutex;
```